<img src="img/panda.jpeg">

**Pandas** - это библиотека Python, предоставляющая широкие возможности для анализа данных. С ее помощью очень удобно загружать, обрабатывать и анализировать табличные данные с помощью SQL-подобных запросов.

Основными структурами данных в Pandas являются классы Series и DataFrame. Первый из них представляет собой одномерный индексированный массив данных некоторого фиксированного типа. Второй - это двухмерная структура данных, представляющая собой таблицу, каждый столбец которой содержит данные одного типа. Можно представлять её как словарь объектов типа Series. Структура DataFrame отлично подходит для представления реальных данных: строки соответствуют признаковым описаниям отдельных объектов, а столбцы соответствуют признакам.

<img src="img/series.PNG">

# Создание DataFrame

In [1]:
import pandas as pd

In [2]:
data = {
    'apples': [3, 2, 0, 1], 
    'oranges': [0, 3, 7, 2]
}

In [3]:
purchases = pd.DataFrame(data)

In [4]:
purchases

apples  oranges
0       3        0
1       2        3
2       0        7
3       1        2

In [5]:
purchases = pd.DataFrame(data, index=['June', 'Robert', 'Lily', 'David'])

purchases

apples  oranges
June         3        0
Robert       2        3
Lily         0        7
David        1        2

In [6]:
df = pd.DataFrame({'id': [1, 2, 3], 'name': ['Harry Potter', 'Ron Weasley', 'Hermione Granger']})

df

id              name
0   1      Harry Potter
1   2       Ron Weasley
2   3  Hermione Granger

# Чтение из файла и первичный анализ

In [7]:
df = pd.read_csv('data/titanic.csv',
                  index_col='PassengerId')

https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html na_values, keep_default_na

In [8]:
df.head()

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  
PassengerId                                                          
1                1      0         A/5 21171   7.2500   NaN        S  
2                1      0          PC 17599  71.2833   C85        C  
3                0      0  STON/O2. 3101282   7.9250   NaN        S  
4                1      0            113803  53.1000  C123        S  
5                0      0            373450   8.0500   NaN        S

In [9]:
df

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   
...               ...     ...   
887                 0       2   
888                 1       1   
889                 0       3   
890                 1       1   
891                 0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   
...                                                        ...     ...   ...   
887                                      Montvila, Rev. Juozas    male  27.0   
888                               Graham, Miss. Margaret Edith  female  19.0   
889                   Johnston, Miss. Catherine Helen "Carrie"  female   NaN   
890                                      Behr, Mr. Karl Howell    male  26.0   
891                                        Dooley, Mr. Patrick    male  32.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  
PassengerId                                                          
1                1      0         A/5 21171   7.2500   NaN        S  
2                1      0          PC 17599  71.2833   C85        C  
3                0      0  STON/O2. 3101282   7.9250   NaN        S  
4                1      0            113803  53.1000  C123        S  
5                0      0            373450   8.0500   NaN        S  
...            ...    ...               ...      ...   ...      ...  
887              0      0            211536  13.0000   NaN        S  
888              0      0            112053  30.0000   B42        S  
889              1      2        W./C. 6607  23.4500   NaN        S  
890              0      0            111369  30.0000  C148        C  
891              0      0            370376   7.7500   NaN        Q  

[891 rows x 11 columns]

In [10]:
df.head(7)

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   
6                   0       3   
7                   0       1   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   
6                                             Moran, Mr. James    male   NaN   
7                                      McCarthy, Mr. Timothy J    male  54.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  
PassengerId                                                          
1                1      0         A/5 21171   7.2500   NaN        S  
2                1      0          PC 17599  71.2833   C85        C  
3                0      0  STON/O2. 3101282   7.9250   NaN        S  
4                1      0            113803  53.1000  C123        S  
5                0      0            373450   8.0500   NaN        S  
6                0      0            330877   8.4583   NaN        Q  
7                0      0             17463  51.8625   E46        S

https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_sql_query.html 

  

In [11]:
df.shape, df.columns

((891, 11),
 Index(['Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch', 'Ticket',
        'Fare', 'Cabin', 'Embarked'],
       dtype='object'))

Для извлечения отдельного столбца можно использовать конструкцию вида DataFrame['Name'].

In [12]:
df['Age'][0:4]

PassengerId
1    22.0
2    38.0
3    26.0
4    35.0
Name: Age, dtype: float64

In [13]:
df['Age'][1:4]

PassengerId
2    38.0
3    26.0
4    35.0
Name: Age, dtype: float64

Для категориальных (тип object) и булевых (тип bool) признаков можно воспользоваться методом value_counts
https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.astype.html

In [14]:
df.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
Int64Index: 891 entries, 1 to 891
Data columns (total 11 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Survived  891 non-null    int64  
 1   Pclass    891 non-null    int64  
 2   Name      891 non-null    object 
 3   Sex       891 non-null    object 
 4   Age       714 non-null    float64
 5   SibSp     891 non-null    int64  
 6   Parch     891 non-null    int64  
 7   Ticket    891 non-null    object 
 8   Fare      891 non-null    float64
 9   Cabin     204 non-null    object 
 10  Embarked  889 non-null    object 
dtypes: float64(2), int64(4), object(5)
memory usage: 318.4 KB


In [15]:
df['Sex'].value_counts()

male      577
female    314
Name: Sex, dtype: int64

# Сортировка

In [16]:
df.sort_values(by='Age').head()

Survived  Pclass                             Name     Sex   Age  \
PassengerId                                                                    
804                 1       3  Thomas, Master. Assad Alexander    male  0.42   
756                 1       2        Hamalainen, Master. Viljo    male  0.67   
645                 1       3           Baclini, Miss. Eugenie  female  0.75   
470                 1       3    Baclini, Miss. Helene Barbara  female  0.75   
79                  1       2    Caldwell, Master. Alden Gates    male  0.83   

             SibSp  Parch  Ticket     Fare Cabin Embarked  
PassengerId                                                
804              0      1    2625   8.5167   NaN        C  
756              1      1  250649  14.5000   NaN        S  
645              2      1    2666  19.2583   NaN        C  
470              2      1    2666  19.2583   NaN        C  
79               0      2  248738  29.0000   NaN        S

In [17]:
 df.sort_values(['Age', 'Sex'], ascending=False).head(20)

Survived  Pclass                                       Name  \
PassengerId                                                                
631                 1       1       Barkworth, Mr. Algernon Henry Wilson   
852                 0       3                        Svensson, Mr. Johan   
97                  0       1                  Goldschmidt, Mr. George B   
494                 0       1                    Artagaveytia, Mr. Ramon   
117                 0       3                       Connors, Mr. Patrick   
673                 0       2                Mitchell, Mr. Henry Michael   
746                 0       1               Crosby, Capt. Edward Gifford   
34                  0       2                      Wheadon, Mr. Edward H   
55                  0       1             Ostby, Mr. Engelhart Cornelius   
281                 0       3                           Duane, Mr. Frank   
457                 0       1                  Millet, Mr. Francis Davis   
439                 0       1                          Fortune, Mr. Mark   
546                 0       1               Nicholson, Mr. Arthur Ernest   
276                 1       1          Andrews, Miss. Kornelia Theodosia   
484                 1       3                     Turkula, Mrs. (Hedwig)   
253                 0       1                  Stead, Mr. William Thomas   
556                 0       1                         Wright, Mr. George   
571                 1       2                         Harris, Mr. George   
830                 1       1  Stone, Mrs. George Nelson (Martha Evelyn)   
171                 0       1                  Van der hoef, Mr. Wyckoff   

                Sex   Age  SibSp  Parch       Ticket      Fare        Cabin  \
PassengerId                                                                   
631            male  80.0      0      0        27042   30.0000          A23   
852            male  74.0      0      0       347060    7.7750          NaN   
97             male  71.0      0      0     PC 17754   34.6542           A5   
494            male  71.0      0      0     PC 17609   49.5042          NaN   
117            male  70.5      0      0       370369    7.7500          NaN   
673            male  70.0      0      0   C.A. 24580   10.5000          NaN   
746            male  70.0      1      1    WE/P 5735   71.0000          B22   
34             male  66.0      0      0   C.A. 24579   10.5000          NaN   
55             male  65.0      0      1       113509   61.9792          B30   
281            male  65.0      0      0       336439    7.7500          NaN   
457            male  65.0      0      0        13509   26.5500          E38   
439            male  64.0      1      4        19950  263.0000  C23 C25 C27   
546            male  64.0      0      0          693   26.0000          NaN   
276          female  63.0      1      0        13502   77.9583           D7   
484          female  63.0      0      0         4134    9.5875          NaN   
253            male  62.0      0      0       113514   26.5500          C87   
556            male  62.0      0      0       113807   26.5500          NaN   
571            male  62.0      0      0  S.W./PP 752   10.5000          NaN   
830          female  62.0      0      0       113572   80.0000          B28   
171            male  61.0      0      0       111240   33.5000          B19   

            Embarked  
PassengerId           
631                S  
852                S  
97                 C  
494                C  
117                Q  
673                S  
746                S  
34                 S  
55                 C  
281                Q  
457                S  
439                S  
546                S  
276                S  
484                S  
253                S  
556                S  
571                S  
830              NaN  
171                S

In [18]:
df.sort_values(by=['Survived', 'Fare'],
        ascending=[True, False]).head()

Survived  Pclass                            Name   Sex   Age  \
PassengerId                                                                 
28                  0       1  Fortune, Mr. Charles Alexander  male  19.0   
439                 0       1               Fortune, Mr. Mark  male  64.0   
119                 0       1        Baxter, Mr. Quigg Edmond  male  24.0   
558                 0       1             Robbins, Mr. Victor  male   NaN   
528                 0       1              Farthing, Mr. John  male   NaN   

             SibSp  Parch    Ticket      Fare        Cabin Embarked  
PassengerId                                                          
28               3      2     19950  263.0000  C23 C25 C27        S  
439              1      4     19950  263.0000  C23 C25 C27        S  
119              0      1  PC 17558  247.5208      B58 B60        C  
558              0      0  PC 17757  227.5250          NaN        C  
528              0      0  PC 17483  221.7792          C95        S

# Извлечение данных

Очень удобной является логическая индексация DataFrame по одному столбцу. Выглядит она следующим образом: df[P(df['Name'])], где P - это некоторое логическое условие, проверяемое для каждого элемента столбца Name. Итогом такой индексации является DataFrame, состоящий только из строк, удовлетворяющих условию P по столбцу Name. Воспользуемся этим для ответа на вопрос:

** Найти пассажиров, которые сели в Cherbourg (Embarked=C) и заплатили более 200 у.е. за билет (fare > 200).**

In [19]:
df[(df['Embarked'] == 'C') & (df['Fare'] > 200)]

Survived  Pclass  \
PassengerId                     
119                 0       1   
259                 1       1   
300                 1       1   
312                 1       1   
378                 0       1   
381                 1       1   
558                 0       1   
680                 1       1   
701                 1       1   
717                 1       1   
738                 1       1   
743                 1       1   

                                                          Name     Sex   Age  \
PassengerId                                                                    
119                                   Baxter, Mr. Quigg Edmond    male  24.0   
259                                           Ward, Miss. Anna  female  35.0   
300            Baxter, Mrs. James (Helene DeLaudeniere Chaput)  female  50.0   
312                                 Ryerson, Miss. Emily Borie  female  18.0   
378                                  Widener, Mr. Harry Elkins    male  27.0   
381                                      Bidois, Miss. Rosalie  female  42.0   
558                                        Robbins, Mr. Victor    male   NaN   
680                         Cardeza, Mr. Thomas Drake Martinez    male  36.0   
701          Astor, Mrs. John Jacob (Madeleine Talmadge Force)  female  18.0   
717                              Endres, Miss. Caroline Louise  female  38.0   
738                                     Lesurer, Mr. Gustave J    male  35.0   
743                      Ryerson, Miss. Susan Parker "Suzette"  female  21.0   

             SibSp  Parch    Ticket      Fare            Cabin Embarked  
PassengerId                                                              
119              0      1  PC 17558  247.5208          B58 B60        C  
259              0      0  PC 17755  512.3292              NaN        C  
300              0      1  PC 17558  247.5208          B58 B60        C  
312              2      2  PC 17608  262.3750  B57 B59 B63 B66        C  
378              0      2    113503  211.5000              C82        C  
381              0      0  PC 17757  227.5250              NaN        C  
558              0      0  PC 17757  227.5250              NaN        C  
680              0      1  PC 17755  512.3292      B51 B53 B55        C  
701              1      0  PC 17757  227.5250          C62 C64        C  
717              0      0  PC 17757  227.5250              C45        C  
738              0      0  PC 17755  512.3292             B101        C  
743              2      2  PC 17608  262.3750  B57 B59 B63 B66        C

Можно отсортировать этих людей по убыванию платы за билет.

In [20]:
df[(df['Embarked'] == 'C') & (df['Fare'] > 200)].sort_values('Fare', ascending=False).head()

Survived  Pclass                                   Name     Sex  \
PassengerId                                                                    
259                 1       1                       Ward, Miss. Anna  female   
680                 1       1     Cardeza, Mr. Thomas Drake Martinez    male   
738                 1       1                 Lesurer, Mr. Gustave J    male   
312                 1       1             Ryerson, Miss. Emily Borie  female   
743                 1       1  Ryerson, Miss. Susan Parker "Suzette"  female   

              Age  SibSp  Parch    Ticket      Fare            Cabin Embarked  
PassengerId                                                                    
259          35.0      0      0  PC 17755  512.3292              NaN        C  
680          36.0      0      1  PC 17755  512.3292      B51 B53 B55        C  
738          35.0      0      0  PC 17755  512.3292             B101        C  
312          18.0      2      2  PC 17608  262.3750  B57 B59 B63 B66        C  
743          21.0      2      2  PC 17608  262.3750  B57 B59 B63 B66        C

Датафреймы можно индексировать как по названию столбца или строки, так и по порядковому номеру. Для индексации **по названию** используется метод **`loc`**, **по номеру** — **`iloc`**.

В первом случае мы говорим _«передай нам значения для id строк от 0 до 5 и для столбцов от Name до Age»_, а во втором — _«передай нам значения первых пяти строк в первых трёх столбцах»_. 

В случае `iloc` срез работает как обычно, однако в случае `loc` учитываются и начало, и конец среза. Да, неудобно, да, вызывает путаницу.

In [21]:
df.loc[0:5, 'Name':'Age']

Name     Sex   Age
PassengerId                                                                 
1                                      Braund, Mr. Owen Harris    male  22.0
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0
3                                       Heikkinen, Miss. Laina  female  26.0
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0
5                                     Allen, Mr. William Henry    male  35.0

In [22]:
df.iloc[0:5, :3]

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   

                                                          Name  
PassengerId                                                     
1                                      Braund, Mr. Owen Harris  
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  
3                                       Heikkinen, Miss. Laina  
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  
5                                     Allen, Mr. William Henry

Если нам нужна первая или последняя строчка датафрейма, пользуемся конструкцией df[:1] или df[-1:]

In [23]:
df[-1:]

Survived  Pclass                 Name   Sex   Age  SibSp  Parch  \
PassengerId                                                                    
891                 0       3  Dooley, Mr. Patrick  male  32.0      0      0   

             Ticket  Fare Cabin Embarked  
PassengerId                               
891          370376  7.75   NaN        Q

# Применение функций: apply, map и др.

**Применение функции к каждому столбцу:**

In [24]:
def age_category(age):
    '''
    < 30 -> 1
    >= 30, <55 -> 2
    >= 55 -> 3
    '''
    if age < 30:
        return 1
    elif age < 55:
        return 2
    else:
        return 3

In [25]:
age_categories = [age_category(age) for age in df.Age]

In [26]:
df['Age_category'] = age_categories
df.head()

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  \
PassengerId                                                           
1                1      0         A/5 21171   7.2500   NaN        S   
2                1      0          PC 17599  71.2833   C85        C   
3                0      0  STON/O2. 3101282   7.9250   NaN        S   
4                1      0            113803  53.1000  C123        S   
5                0      0            373450   8.0500   NaN        S   

             Age_category  
PassengerId                
1                       1  
2                       2  
3                       1  
4                       2  
5                       2

In [27]:
df['Age_category'] = df['Age'].apply(age_category)
df.head()

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  \
PassengerId                                                           
1                1      0         A/5 21171   7.2500   NaN        S   
2                1      0          PC 17599  71.2833   C85        C   
3                0      0  STON/O2. 3101282   7.9250   NaN        S   
4                1      0            113803  53.1000  C123        S   
5                0      0            373450   8.0500   NaN        S   

             Age_category  
PassengerId                
1                       1  
2                       2  
3                       1  
4                       2  
5                       2

In [28]:
df = df.drop('Cabin', axis=1)

In [29]:
df.head()

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   

             SibSp  Parch            Ticket     Fare Embarked  Age_category  
PassengerId                                                                  
1                1      0         A/5 21171   7.2500        S             1  
2                1      0          PC 17599  71.2833        C             2  
3                0      0  STON/O2. 3101282   7.9250        S             1  
4                1      0            113803  53.1000        S             2  
5                0      0            373450   8.0500        S             2

In [30]:
df = df.rename(columns=lambda x: x.upper())
df.head()

SURVIVED  PCLASS  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   

                                                          NAME     SEX   AGE  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   

             SIBSP  PARCH            TICKET     FARE EMBARKED  AGE_CATEGORY  
PassengerId                                                                  
1                1      0         A/5 21171   7.2500        S             1  
2                1      0          PC 17599  71.2833        C             2  
3                0      0  STON/O2. 3101282   7.9250        S             1  
4                1      0            113803  53.1000        S             2  
5                0      0            373450   8.0500        S             2

In [31]:
df['NAME'] = df['NAME'].apply(lambda x: x.upper())
df.head() 

SURVIVED  PCLASS  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   

                                                          NAME     SEX   AGE  \
PassengerId                                                                    
1                                      BRAUND, MR. OWEN HARRIS    male  22.0   
2            CUMINGS, MRS. JOHN BRADLEY (FLORENCE BRIGGS TH...  female  38.0   
3                                       HEIKKINEN, MISS. LAINA  female  26.0   
4                 FUTRELLE, MRS. JACQUES HEATH (LILY MAY PEEL)  female  35.0   
5                                     ALLEN, MR. WILLIAM HENRY    male  35.0   

             SIBSP  PARCH            TICKET     FARE EMBARKED  AGE_CATEGORY  
PassengerId                                                                  
1                1      0         A/5 21171   7.2500        S             1  
2                1      0          PC 17599  71.2833        C             2  
3                0      0  STON/O2. 3101282   7.9250        S             1  
4                1      0            113803  53.1000        S             2  
5                0      0            373450   8.0500        S             2

Метод `map` можно использовать и для **замены значений в колонке**, передав ему в качестве аргумента словарь вида `{old_value: new_value}`:

In [32]:
d = {'female' : 1, 'male' : 0}
df['SEX'] = df['SEX'].map(d)
df.head()

SURVIVED  PCLASS  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   

                                                          NAME  SEX   AGE  \
PassengerId                                                                 
1                                      BRAUND, MR. OWEN HARRIS    0  22.0   
2            CUMINGS, MRS. JOHN BRADLEY (FLORENCE BRIGGS TH...    1  38.0   
3                                       HEIKKINEN, MISS. LAINA    1  26.0   
4                 FUTRELLE, MRS. JACQUES HEATH (LILY MAY PEEL)    1  35.0   
5                                     ALLEN, MR. WILLIAM HENRY    0  35.0   

             SIBSP  PARCH            TICKET     FARE EMBARKED  AGE_CATEGORY  
PassengerId                                                                  
1                1      0         A/5 21171   7.2500        S             1  
2                1      0          PC 17599  71.2833        C             2  
3                0      0  STON/O2. 3101282   7.9250        S             1  
4                1      0            113803  53.1000        S             2  
5                0      0            373450   8.0500        S             2

In [33]:
d = {1 : 'female', 0: 'male'}
df = df.replace({'SEX': d})
df.head()

SURVIVED  PCLASS  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   

                                                          NAME     SEX   AGE  \
PassengerId                                                                    
1                                      BRAUND, MR. OWEN HARRIS    male  22.0   
2            CUMINGS, MRS. JOHN BRADLEY (FLORENCE BRIGGS TH...  female  38.0   
3                                       HEIKKINEN, MISS. LAINA  female  26.0   
4                 FUTRELLE, MRS. JACQUES HEATH (LILY MAY PEEL)  female  35.0   
5                                     ALLEN, MR. WILLIAM HENRY    male  35.0   

             SIBSP  PARCH            TICKET     FARE EMBARKED  AGE_CATEGORY  
PassengerId                                                                  
1                1      0         A/5 21171   7.2500        S             1  
2                1      0          PC 17599  71.2833        C             2  
3                0      0  STON/O2. 3101282   7.9250        S             1  
4                1      0            113803  53.1000        S             2  
5                0      0            373450   8.0500        S             2

In [34]:
df = pd.read_csv('data/titanic.csv',
                  index_col='PassengerId')

### Группировка данных

В общем случае группировка данных в Pandas выглядит следующим образом:

```
df.groupby(by=grouping_columns)[columns_to_show].function()
```

1. К датафрейму применяется метод **`groupby`**, который разделяет данные по `grouping_columns` – признаку или набору признаков.
3. Индексируем по нужным нам столбцам (`columns_to_show`). 
2. К полученным группам применяется функция или несколько функций.

In [35]:
df.groupby(['Survived', 'Sex'])['Survived'].count()

Survived  Sex   
0         female     81
          male      468
1         female    233
          male      109
Name: Survived, dtype: int64

# Сводные таблицы

Допустим, мы хотим посмотреть, как наблюдения в нашей выборке распределены в контексте двух признаков. Для этого мы можем построить **таблицу сопряженности**, воспользовавшись методом **`crosstab`**:

In [36]:
pd.crosstab(df['Pclass'], df['Survived'])

Survived    0    1
Pclass            
1          80  136
2          97   87
3         372  119

* `values` – список переменных, по которым требуется рассчитать нужные статистики,
* `index` – список переменных, по которым нужно сгруппировать данные,
* `aggfunc` — то, что нам, собственно, нужно посчитать по группам — сумму, среднее, максимум, минимум или что-то ещё.

In [37]:
pd.crosstab(df['Pclass'], df['Survived'], values=df['Age'], aggfunc='max')

Survived     0     1
Pclass              
1         71.0  80.0
2         70.0  62.0
3         74.0  63.0

# Pandas vs SQL

In [38]:
airports = pd.read_csv('data/airports.csv')
airport_freq = pd.read_csv('data/airport-frequencies.csv')
runways = pd.read_csv('data/runways.csv')

In [39]:
airports.head()

id ident           type                                name  \
0    6523   00A       heliport                   Total Rf Heliport   
1  323361  00AA  small_airport                Aero B Ranch Airport   
2    6524  00AK  small_airport                        Lowell Field   
3    6525  00AL  small_airport                        Epps Airpark   
4    6526  00AR         closed  Newport Hospital & Clinic Heliport   

   latitude_deg  longitude_deg  elevation_ft continent iso_country iso_region  \
0     40.070801     -74.933601          11.0       NaN          US      US-PA   
1     38.704022    -101.473911        3435.0       NaN          US      US-KS   
2     59.949200    -151.695999         450.0       NaN          US      US-AK   
3     34.864799     -86.770302         820.0       NaN          US      US-AL   
4     35.608700     -91.254898         237.0       NaN          US      US-AR   

   municipality scheduled_service gps_code iata_code local_code home_link  \
0      Bensalem                no      00A       NaN        00A       NaN   
1         Leoti                no     00AA       NaN       00AA       NaN   
2  Anchor Point                no     00AK       NaN       00AK       NaN   
3       Harvest                no     00AL       NaN       00AL       NaN   
4       Newport                no      NaN       NaN        NaN       NaN   

  wikipedia_link keywords  
0            NaN      NaN  
1            NaN      NaN  
2            NaN      NaN  
3            NaN      NaN  
4            NaN     00AR

## SELECT, WHERE, DISTINCT, LIMIT

select id from airports where ident = 'KLAX'

In [40]:
airports[airports.ident == 'KLAX']['id']

28067    3632
Name: id, dtype: int64

select distinct type from airport

In [41]:
airports.type.unique()

array(['heliport', 'small_airport', 'closed', 'seaplane_base',
       'balloonport', 'medium_airport', 'large_airport'], dtype=object)

## SELECT with multiple conditions

select * from airports where iso_region = 'US-CA' and type = 'seaplane_base'

In [42]:
airports[(airports.iso_region == 'US-CA') & (airports.type == 'seaplane_base')].head() 

id ident           type                                 name  \
979     7436   0O0  seaplane_base     San Luis Reservoir Seaplane Base   
2497    8877  22CA  seaplane_base       Commodore Center Seaplane Base   
6258   12298  5CA9  seaplane_base  Konocti  - Clear Lake Seaplane Base   
13051  16514   C39  seaplane_base            Folsom Lake Seaplane Base   
15378  16830  CN20  seaplane_base        Ferndale Resort Seaplane Base   

       latitude_deg  longitude_deg  elevation_ft continent iso_country  \
979       37.058300    -121.125999         544.0       NaN          US   
2497      37.879902    -122.514000           NaN       NaN          US   
6258      38.977699    -122.718002        1326.0       NaN          US   
13051     38.707199    -121.133003         466.0       NaN          US   
15378     39.002998    -122.796997        1326.0       NaN          US   

      iso_region municipality scheduled_service gps_code iata_code local_code  \
979        US-CA    Los Banos                no      0O0       NaN        0O0   
2497       US-CA    Sausalito                no     22CA       NaN       22CA   
6258       US-CA  Kelseyville                no     5CA9       NaN       5CA9   
13051      US-CA       Folsom                no      C39       NaN        C39   
15378      US-CA  Kelseyville                no     CN20       NaN       CN20   

      home_link wikipedia_link keywords  
979         NaN            NaN      NaN  
2497        NaN            NaN      NaN  
6258        NaN            NaN      NaN  
13051       NaN            NaN      NaN  
15378       NaN            NaN      NaN

select ident, name, municipality from airports where iso_region = 'US-CA' and type = 'large_airport'

In [43]:
airports[(airports.iso_region == 'US-CA') & (airports.type == 'large_airport')][['ident', 'name', 'municipality']]

ident                                             name   municipality
26439  KBAB                             Beale Air Force Base     Marysville
27277  KEDW                           Edwards Air Force Base        Edwards
28067  KLAX                Los Angeles International Airport    Los Angeles
28594  KOAK       Metropolitan Oakland International Airport        Oakland
28650  KONT                    Ontario International Airport        Ontario
30170  KSAN                  San Diego International Airport      San Diego
30210  KSFO              San Francisco International Airport  San Francisco
30231  KSJC  Norman Y. Mineta San Jose International Airport       San Jose
30253  KSMF                 Sacramento International Airport     Sacramento
30259  KSNA         John Wayne Airport-Orange County Airport      Santa Ana
30308  KSUU                            Travis Air Force Base      Fairfield
30541  KVBG                        Vandenberg Air Force Base         Lompoc

## ORDER BY

select * from airport_freq where airport_ident = 'KLAX' order by type

In [44]:
airport_freq[airport_freq.airport_ident == 'KLAX'].sort_values('type')

id  airport_ref airport_ident  type description  frequency_mhz
11916  60767         3632          KLAX   APP   SOCAL APP          36.07
11917  60766         3632          KLAX   APP   SOCAL APP         124.30
11918  60768         3632          KLAX  ATIS        ATIS         133.80
11919  60769         3632          KLAX   CLD    CLNC DEL         121.40
11920  60770         3632          KLAX   DEP   SOCAL DEP         124.30
11921  60771         3632          KLAX   GND         GND         121.65
11922  60772         3632          KLAX  MISC          CG          34.50
11923  60773         3632          KLAX  MISC          CG         898.40
11924  60774         3632          KLAX   OPS          AF          37.22
11925  60775         3632          KLAX   TWR         TWR         119.80
11926  60776         3632          KLAX  UNIC      UNICOM         122.95

select * from airport_freq where airport_ident = 'KLAX' order by type desc

In [45]:
airport_freq[airport_freq.airport_ident == 'KLAX'].sort_values('type', ascending=False)

id  airport_ref airport_ident  type description  frequency_mhz
11926  60776         3632          KLAX  UNIC      UNICOM         122.95
11925  60775         3632          KLAX   TWR         TWR         119.80
11924  60774         3632          KLAX   OPS          AF          37.22
11922  60772         3632          KLAX  MISC          CG          34.50
11923  60773         3632          KLAX  MISC          CG         898.40
11921  60771         3632          KLAX   GND         GND         121.65
11920  60770         3632          KLAX   DEP   SOCAL DEP         124.30
11919  60769         3632          KLAX   CLD    CLNC DEL         121.40
11918  60768         3632          KLAX  ATIS        ATIS         133.80
11916  60767         3632          KLAX   APP   SOCAL APP          36.07
11917  60766         3632          KLAX   APP   SOCAL APP         124.30

# IN… NOT IN

select * from airports where type in ('heliport', 'balloonport')

In [46]:
airports[airports.type.isin(['heliport', 'balloonport'])]

id    ident      type                             name  \
0        6523      00A  heliport                Total Rf Heliport   
9      322658     00CN  heliport  Kitchen Creek Helibase Heliport   
12       6532     00FD  heliport               Ringhaver Heliport   
15       6535     00GE  heliport                 Caffrey Heliport   
16       6536     00HI  heliport               Kaupulehu Heliport   
...       ...      ...       ...                              ...   
55221  312611  ZA-0117  heliport                    Pumba Helipad   
55223  312624  ZA-0119  heliport               Cape Town Heliport   
55244  318475  ZA-0140  heliport        Kuruman Hospital Heliport   
55340  301278      ZIZ  heliport                 Zamzama Heliport   
55601   46378  ZZ-0001  heliport                  Sealand Helipad   

       latitude_deg  longitude_deg  elevation_ft continent iso_country  \
0         40.070801     -74.933601          11.0       NaN          US   
9         32.727374    -116.459742        3350.0       NaN          US   
12        28.846600     -82.345398          25.0       NaN          US   
15        33.884201     -84.733902         957.0       NaN          US   
16        19.832715    -155.980233          43.0       NaN          US   
...             ...            ...           ...       ...         ...   
55221    -33.387799      26.410595           NaN        AF          ZA   
55223    -33.901018      18.425936           NaN        AF          ZA   
55244    -27.459904      23.443762           NaN        AF          ZA   
55340     26.710944      67.667250         128.0        AS          PK   
55601     51.894444       1.482500          40.0        EU          GB   

      iso_region        municipality scheduled_service gps_code iata_code  \
0          US-PA            Bensalem                no      00A       NaN   
9          US-CA         Pine Valley                no     00CN       NaN   
12         US-FL           Riverview                no     00FD       NaN   
15         US-GA               Hiram                no     00GE       NaN   
16         US-HI         Kailua/Kona                no     00HI       NaN   
...          ...                 ...               ...      ...       ...   
55221      ZA-EC  Nelson Mandela Bay                no      NaN       NaN   
55223      ZA-WC                 NaN                no      NaN       NaN   
55244      ZA-NC             Kuruman                no      NaN       NaN   
55340      PK-SD   Zamzama Gas Field                no      NaN       ZIZ   
55601     GB-ENG             Sealand                no      NaN       NaN   

      local_code                                    home_link  \
0            00A                                          NaN   
9           00CN                                          NaN   
12          00FD                                          NaN   
15          00GE                                          NaN   
16          00HI                                          NaN   
...          ...                                          ...   
55221        NaN  http://www.pumbagamereserve.co.za/amenities   
55223        NaN                                          NaN   
55244        NaN                                          NaN   
55340        NaN                                          NaN   
55601        NaN                   http://www.sealandgov.org/   

                                          wikipedia_link              keywords  
0                                                    NaN                   NaN  
9                                                    NaN                   NaN  
12                                                   NaN                   NaN  
15                                                   NaN                   NaN  
16                                                   NaN                   NaN  
...                                                  ...                   ...  
55221               

In [47]:
# select * from airports where type not in ('heliport', 'balloonport')
airports[~airports.type.isin(['heliport', 'balloonport'])]

id    ident            type                                name  \
1      323361     00AA   small_airport                Aero B Ranch Airport   
2        6524     00AK   small_airport                        Lowell Field   
3        6525     00AL   small_airport                        Epps Airpark   
4        6526     00AR          closed  Newport Hospital & Clinic Heliport   
5      322127     00AS   small_airport                      Fulton Airport   
...       ...      ...             ...                                 ...   
55598   27244     ZYYJ  medium_airport         Yanji Chaoyangchuan Airport   
55599  317861     ZYYK  medium_airport               Yingkou Lanqi Airport   
55600   32753     ZYYY  medium_airport             Shenyang Dongta Airport   
55602  307326  ZZ-0002   small_airport           Glorioso Islands Airstrip   
55603  313629     ZZZZ   small_airport              Satsuma Iōjima Airport   

       latitude_deg  longitude_deg  elevation_ft continent iso_country  \
1         38.704022    -101.473911        3435.0       NaN          US   
2         59.949200    -151.695999         450.0       NaN          US   
3         34.864799     -86.770302         820.0       NaN          US   
4         35.608700     -91.254898         237.0       NaN          US   
5         34.942803     -97.818019        1100.0       NaN          US   
...             ...            ...           ...       ...         ...   
55598     42.882801     129.451004         624.0        AS          CN   
55599     40.542524     122.358600           0.0        AS          CN   
55600     41.784401     123.496002           NaN        AS          CN   
55602    -11.584278      47.296389          11.0        AF          TF   
55603     30.784722     130.270556         338.0        AS          JP   

      iso_region      municipality scheduled_service gps_code iata_code  \
1          US-KS             Leoti                no     00AA       NaN   
2          US-AK      Anchor Point                no     00AK       NaN   
3          US-AL           Harvest                no     00AL       NaN   
4          US-AR           Newport                no      NaN       NaN   
5          US-OK              Alex                no     00AS       NaN   
...          ...               ...               ...      ...       ...   
55598      CN-22             Yanji               yes     ZYYJ       YNJ   
55599      CN-21           Yingkou               yes     ZYYK       YKH   
55600      CN-21          Shenyang                no     ZYYY       NaN   
55602     TF-U-A  Grande Glorieuse                no      NaN       NaN   
55603      JP-46      Mishima-Mura                no     RJX7       NaN   

      local_code home_link                                     wikipedia_link  \
1           00AA       NaN                                                NaN   
2           00AK       NaN                                                NaN   
3           00AL       NaN                                                NaN   
4            NaN       NaN                                                NaN   
5           00AS       NaN                                                NaN   
...          ...       ...                                                ...   
55598        NaN       NaN  https://en.wikipedia.org/wiki/Yanji_Chaoyangch...   
55599        NaN       NaN  https://en.wikipedia.org/wiki/Yingkou_Lanqi_Ai...   
55600        NaN       NaN                                                NaN   
55602        NaN       NaN                                                NaN   
55603        NaN       NaN  http://wikimapia.org/6705190/Satsuma-Iwo-jima-...   

                   keywords  
1                       NaN  
2                       NaN  
3                       NaN  
4                      00AR  
5                       NaN  
...                     ...  
55598                   NaN  
55599                   NaN  
55600                   NaN  
55602      

# GROUP BY, COUNT, ORDER BY

select iso_country, type, count(*)  from airports group by iso_country, type order by iso_country, type

In [48]:
airports.groupby(['iso_country', 'type']).size()

iso_country  type          
AD           heliport            2
AE           closed              1
             heliport           21
             large_airport       4
             medium_airport      7
                              ... 
ZM           small_airport      94
ZW           closed              2
             large_airport       1
             medium_airport      8
             small_airport     128
Length: 841, dtype: int64

select iso_country, type, count() from airports group by iso_country, type order by iso_country, count() desc

In [49]:
airports.groupby(['iso_country', 'type']).size().to_frame('size').reset_index().sort_values(['iso_country', 'size'], ascending=[True, False])

iso_country            type  size
0            AD        heliport     2
2            AE        heliport    21
6            AE   small_airport    15
4            AE  medium_airport     7
3            AE   large_airport     4
..          ...             ...   ...
834          ZM   large_airport     1
840          ZW   small_airport   128
839          ZW  medium_airport     8
837          ZW          closed     2
838          ZW   large_airport     1

[841 rows x 3 columns]

# HAVING

select type, count() from airports where iso_country = 'US' group by type having count() > 1000 order by count(*) desc

In [50]:
airports[airports.iso_country == 'US'].groupby('type').filter(lambda g: len(g) > 1000).groupby('type').size().sort_values(ascending=False)

type
small_airport    13679
heliport          6260
closed            1458
dtype: int64

# Aggregate functions (MIN, MAX, MEAN)


select max(length_ft), min(length_ft), avg(length_ft), median(length_ft) from runways

In [51]:
runways.agg({'length_ft': ['min', 'max'], 'width_ft': ['max', 'mean', 'median']})

length_ft    width_ft
max      120000.0  9000.00000
mean          NaN   110.61706
median        NaN    75.00000
min           0.0         NaN

## JOIN

### inner (default), left (corresponds to LEFT OUTER in SQL), right (RIGHT OUTER), or outer (FULL OUTER)

select airport_ident, type, description, frequency_mhz from airport_freq join airports on airport_freq.airport_ref = airports.id where airports.ident = 'KLAX'

In [52]:
airport_freq.merge(airports[airports.ident == 'KLAX'][['id']], left_on='airport_ref', right_on='id', how='inner')[['airport_ident', 'type', 'description', 'frequency_mhz']]

airport_ident  type description  frequency_mhz
0           KLAX   APP   SOCAL APP          36.07
1           KLAX   APP   SOCAL APP         124.30
2           KLAX  ATIS        ATIS         133.80
3           KLAX   CLD    CLNC DEL         121.40
4           KLAX   DEP   SOCAL DEP         124.30
5           KLAX   GND         GND         121.65
6           KLAX  MISC          CG          34.50
7           KLAX  MISC          CG         898.40
8           KLAX   OPS          AF          37.22
9           KLAX   TWR         TWR         119.80
10          KLAX  UNIC      UNICOM         122.95

## UNION ALL

pd.concat() - UNION ALL

select name, municipality from airports where ident = 'KLAX' union all select name, municipality from airports where ident = 'KLGB'

In [53]:
pd.concat([airports[airports.ident == 'KLAX'][['name', 'municipality']], airports[airports.ident == 'KLGB'][['name', 'municipality']]])

name municipality
28067     Los Angeles International Airport  Los Angeles
28092  Long Beach /Daugherty Field/ Airport   Long Beach

## ROW_NUMBER() OVER 
Пример для нахождения первой строки для страны (с предварительной сортировкой) 

<code>
    
df = df.sort_values(['SCORE'], ascending=False).drop_duplicates(
    ['CTRY'])
 
SELECT * FROM (
    SELECT CNTRY, 
    SCORE, 
    ...other columns...,
    ROW_NUMBER() OVER (PARTITION BY CTRY ORDER BY SCORE DESC ) RN
    ) F
WHERE F.RN = 1  

## DELETE

In [54]:
airports = airports[airports.ident != 'KLAX']
airports

id    ident            type                                name  \
0        6523      00A        heliport                   Total Rf Heliport   
1      323361     00AA   small_airport                Aero B Ranch Airport   
2        6524     00AK   small_airport                        Lowell Field   
3        6525     00AL   small_airport                        Epps Airpark   
4        6526     00AR          closed  Newport Hospital & Clinic Heliport   
...       ...      ...             ...                                 ...   
55599  317861     ZYYK  medium_airport               Yingkou Lanqi Airport   
55600   32753     ZYYY  medium_airport             Shenyang Dongta Airport   
55601   46378  ZZ-0001        heliport                     Sealand Helipad   
55602  307326  ZZ-0002   small_airport           Glorioso Islands Airstrip   
55603  313629     ZZZZ   small_airport              Satsuma Iōjima Airport   

       latitude_deg  longitude_deg  elevation_ft continent iso_country  \
0         40.070801     -74.933601          11.0       NaN          US   
1         38.704022    -101.473911        3435.0       NaN          US   
2         59.949200    -151.695999         450.0       NaN          US   
3         34.864799     -86.770302         820.0       NaN          US   
4         35.608700     -91.254898         237.0       NaN          US   
...             ...            ...           ...       ...         ...   
55599     40.542524     122.358600           0.0        AS          CN   
55600     41.784401     123.496002           NaN        AS          CN   
55601     51.894444       1.482500          40.0        EU          GB   
55602    -11.584278      47.296389          11.0        AF          TF   
55603     30.784722     130.270556         338.0        AS          JP   

      iso_region      municipality scheduled_service gps_code iata_code  \
0          US-PA          Bensalem                no      00A       NaN   
1          US-KS             Leoti                no     00AA       NaN   
2          US-AK      Anchor Point                no     00AK       NaN   
3          US-AL           Harvest                no     00AL       NaN   
4          US-AR           Newport                no      NaN       NaN   
...          ...               ...               ...      ...       ...   
55599      CN-21           Yingkou               yes     ZYYK       YKH   
55600      CN-21          Shenyang                no     ZYYY       NaN   
55601     GB-ENG           Sealand                no      NaN       NaN   
55602     TF-U-A  Grande Glorieuse                no      NaN       NaN   
55603      JP-46      Mishima-Mura                no     RJX7       NaN   

      local_code                   home_link  \
0            00A                         NaN   
1           00AA                         NaN   
2           00AK                         NaN   
3           00AL                         NaN   
4            NaN                         NaN   
...          ...                         ...   
55599        NaN                         NaN   
55600        NaN                         NaN   
55601        NaN  http://www.sealandgov.org/   
55602        NaN                         NaN   
55603        NaN                         NaN   

                                          wikipedia_link              keywords  
0                                                    NaN                   NaN  
1                                                    NaN                   NaN  
2                                                    NaN                   NaN  
3                                                    NaN                   NaN  
4                                                    NaN                  00AR  
...                                                  ...                   ...  
55599  https://en.wikipedia.org/wiki/Yingkou_Lanqi_Ai...                   NaN  
55600                                                NaN   

Удалилась одна строка

# Упражнения по titanic.csv

**1. Сколько мужчин / женщин находилось на борту?**
- 412 мужчин и 479 женщин
- 314 мужчин и 577 женщин
- 479 мужчин и 412 женщин
- 577 мужчин и 314 женщин

In [55]:
df['Sex'].value_counts()

male      577
female    314
Name: Sex, dtype: int64

**2. Выведите распределение переменной `Pclass` (социально-экономический статус) и это же распределение, только для мужчин / женщин по отдельности. Сколько было мужчин 2-го класса?**
- 104
- 108
- 112
- 125

In [56]:
pd.crosstab(df['Sex'], df['Pclass'])

Pclass    1    2    3
Sex                  
female   94   76  144
male    122  108  347

In [57]:
df.groupby(['Sex', 'Pclass'])['Sex'].count()

Sex     Pclass
female  1          94
        2          76
        3         144
male    1         122
        2         108
        3         347
Name: Sex, dtype: int64

**3. Каковы медиана и стандартное отклонение платежей (`Fare`)? Округлите до 2 десятичных знаков.**
- Медиана – 14.45, стандартное отклонение – 49.69
- Медиана – 15.1, стандартное отклонение – 12.15
- Медиана – 13.15, стандартное отклонение – 35.3
- Медиана – 17.43, стандартное отклонение – 39.1

In [58]:
df['Fare'].describe()

count    891.000000
mean      32.204208
std       49.693429
min        0.000000
25%        7.910400
50%       14.454200
75%       31.000000
max      512.329200
Name: Fare, dtype: float64

In [59]:
df.agg({'Fare': ['median', 'std']})

Fare
median  14.454200
std     49.693429

In [60]:
df['Fare'].median()

14.4542

In [61]:
df['Fare'].std()

49.6934285971809

**4. Правда ли, что люди моложе 30 лет выживали чаще, чем люди старше 60 лет? Каковы доли выживших в обеих группах?**
- 22.7% среди молодых и 40.6% среди старых
- 40.6% среди молодых и 22.7% среди старых
- 35.3% среди молодых и 27.4% среди старых
- 27.4% среди молодых и 35.3% среди старых

In [62]:
df.Survived[df.Age < 30].sum() / df.Survived[df.Age < 30].count()

0.40625

In [63]:
df[(df['Age'] < 30) & df['Survived'] == 1].shape[0] / df[df['Age'] < 30].shape[0]

0.40625

In [64]:
df.Survived[df.Age > 60].sum() / df.Survived[df.Age > 60].count()

0.22727272727272727

In [65]:
df[(df['Age'] > 60) & df['Survived'] == 1].shape[0] / df[df['Age'] > 60].shape[0]

0.22727272727272727

**5. Правда ли, что женщины выживали чаще мужчин? Каковы доли выживших в обеих группах?**
- 30.2% среди мужчин и 46.2% среди женщин
- 35.7% среди мужчин и 74.2% среди женщин
- 21.1% среди мужчин и 46.2% среди женщин
- 18.9% среди мужчин и 74.2% среди женщин

In [66]:
df[(df['Sex'] == 'male') & df['Survived'] == 1].shape[0] / df[df['Sex'] == 'male'].shape[0]

0.18890814558058924

In [67]:
df[(df['Sex'] == 'female') & df['Survived'] == 1].shape[0] / df[df['Sex'] == 'female'].shape[0]

0.7420382165605095

**6. Как отличается средний возраст мужчин / женщин в зависимости от класса обслуживания? Выберите верные утверждения:**
- В среднем мужчины 1-го класса старше 40 лет
- В среднем женщины 1-го класса старше 40 лет
- Мужчины всех классов в среднем старше женщин того же класса
- В среднем люди в 1 классе старше, чем во 2-ом, а те старше представителей 3-го класса

In [68]:
pd.crosstab(df['Pclass'], df['Sex'], values=df['Age'], aggfunc='mean')

Sex        female       male
Pclass                      
1       34.611765  41.281386
2       28.722973  30.740707
3       21.750000  26.507589